src/probability/martingale/convergence.lean

```diff
@@ -0,0 +1,287 @@
+          λ n, le_trans _ (hR' n)⟩),
+      refine lintegral_mono (λ ω, _),
+      rw [ennreal.of_real_le_iff_le_to_real, ennreal.coe_to_real, coe_nnnorm],
+      by_cases hnonneg : 0 ≤ f n ω - a,
+      { rw [lattice_ordered_comm_group.pos_of_nonneg _ hnonneg,
+          real.norm_eq_abs, abs_of_nonneg hnonneg] },
+      { rw lattice_ordered_comm_group.pos_of_nonpos _ (not_le.1 hnonneg).le,
+        exact norm_nonneg _ },
+      { simp only [ne.def, ennreal.coe_ne_top, not_false_iff] } },
+    { simp only [hab, ne.def, ennreal.of_real_eq_zero, sub_nonpos, not_le] } },
+  { simp only [hab, ne.def, ennreal.of_real_eq_zero, sub_nonpos, not_le, true_or]},
+  { simp only [ne.def, ennreal.of_real_ne_top, not_false_iff, true_or] }
+end
+
+lemma submartingale.upcrossings_ae_lt_top [is_finite_measure μ]
+  (hf : submartingale f ℱ μ) (hbdd : ∀ n, snorm (f n) 1 μ ≤ R) :
+  ∀ᵐ ω ∂μ, ∀ a b : ℚ, a < b → upcrossings a b f ω < ∞ :=
+begin
+  simp only [ae_all_iff, eventually_imp_distrib_left],
+  rintro a b hab,
+  exact hf.upcrossings_ae_lt_top' hbdd (rat.cast_lt.2 hab),
+end
+
+/-- An L¹-bounded submartingale converges almost everywhere. -/
+lemma submartingale.exists_ae_tendsto_of_bdd [is_finite_measure μ]
+  (hf : submartingale f ℱ μ) (hbdd : ∀ n, snorm (f n) 1 μ ≤ R) :
+  ∀ᵐ ω ∂μ, ∃ c, tendsto (λ n, f n ω) at_top (𝓝 c) :=
+begin
+  filter_upwards [hf.upcrossings_ae_lt_top hbdd, liminf_at_top_ae_bdd_of_snorm_one_bdd
+    (λ n, (hf.strongly_measurable n).measurable.mono (ℱ.le n) le_rfl) hbdd] with ω h₁ h₂,
+  exact tendsto_of_uncrossing_lt_top h₂ h₁,
+end
+
+lemma submartingale.exists_ae_trim_tendsto_of_bdd [is_finite_measure μ]
+  (hf : submartingale f ℱ μ) (hbdd : ∀ n, snorm (f n) 1 μ ≤ R) :
+  ∀ᵐ ω ∂(μ.trim (Sup_le (λ m ⟨n, hn⟩, hn ▸ ℱ.le _) : (⨆ n, ℱ n) ≤ m0)),
+    ∃ c, tendsto (λ n, f n ω) at_top (𝓝 c) :=
+begin
+  rw [ae_iff, trim_measurable_set_eq],
+  { exact hf.exists_ae_tendsto_of_bdd hbdd },
+  { exact measurable_set.compl (@measurable_set_exists_tendsto _ _ _ _ _ _ (⨆ n, ℱ n) _ _ _ _ _
+    (λ n, ((hf.strongly_measurable n).measurable.mono (le_Sup ⟨n, rfl⟩) le_rfl))) }
+end
+
+section limit
+
+open_locale classical
+
+variables [preorder ι] {E : Type*} [has_zero E] [topological_space E]
+  {𝒢 : filtration ι m0} {g : ι → Ω → E}
+
+/-- Given a process `f` and a filtration `ℱ`, if `f` converges to some `g` almost everywhere and
+`g` is `⨆ n, ℱ n`-measurable, then `limit_process f ℱ` chooses said `g`, else it returns 0.
+
+This definition is used to phrase the a.e. martingale convergence theorem
+`submartingale.ae_tendsto_limit_process` where an L¹-bounded submartingale `f` adapted to `ℱ`
+converges to `limit_process f ℱ μ` `μ`-almost everywhere. -/
+noncomputable
+def limit_process (f : ι → Ω → E) (ℱ : filtration ι m0) (μ : measure Ω) :=
+if h : ∃ g : Ω → E, strongly_measurable[⨆ n, ℱ n] g ∧
+  ∀ᵐ ω ∂μ, tendsto (λ n, f n ω) at_top (𝓝 (g ω)) then classical.some h else 0
+
+lemma limit_process_measurable :
```

Comment:
The order of the terms should be reversed. In the same way that we have a lemma `measurable_exp`, and so on. Moreover, since it mentions strong measurability it should rather be phrased `strongly_measurable_limit_process`.